File: src/models/Lead.ts
import { Schema, model, Document } from 'mongoose';

export interface ILead extends Document {
  lead_id: string;
  name?: string;
  phone_number?: string;
  email?: string;
  aadhar_number?: string;
  source_id?: string;
  p_id?: string;
  created_at: Date;
  
  // Merge metadata
  merged_from: any[];
  sources_seen: string[];
  products_seen: string[];
  
  // AI scoring stub fields
  lead_score: number | null;
  score_reason: string | null;
  
  createdAt: Date;
  updatedAt: Date;
}

const LeadSchema = new Schema<ILead>(
  {
    lead_id: {
      type: String,
      required: true,
      unique: true,
      trim: true
    },
    name: {
      type: String,
      trim: true
    },
    phone_number: {
      type: String,
      trim: true,
      index: true
    },
    email: {
      type: String,
      trim: true,
      lowercase: true,
      index: true
    },
    aadhar_number: {
      type: String,
      trim: true,
      index: true
    },
    source_id: {
      type: String,
      trim: true
    },
    p_id: {
      type: String,
      trim: true
    },
    created_at: {
      type: Date,
      required: true
    },
    
    // Metadata for tracking merges
    merged_from: {
      type: [Array],
      default: []
    },
    sources_seen: {
      type: [String],
      default: []
    },
    products_seen: {
      type: [String],
      default: []
    },
    
    // AI scoring placeholder
    lead_score: {
      type: Number,
      default: null
    },
    score_reason: {
      type: String,
      default: null
    }
  },
  {
    timestamps: true,
    collection: 'leads'
  }
);

// Compound index for deduplication queries
LeadSchema.index(
  { email: 1, phone_number: 1, aadhar_number: 1 },
  { name: 'lead_identifiers_index' }
);

// Index for dashboard queries
LeadSchema.index({ p_id: 1, source_id: 1, created_at: -1 });

export const Lead = model<ILead>('Lead', LeadSchema);

File: bank-leads-java/bank-leads-backend/src/main/java/com/bankleads/bank_leads_backend/model/Lead.java
package com.bankleads.bank_leads_backend.model;

import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;
import org.springframework.data.annotation.CreatedDate;
import org.springframework.data.annotation.Id;
import org.springframework.data.annotation.LastModifiedDate;
import org.springframework.data.mongodb.core.index.CompoundIndex;
import org.springframework.data.mongodb.core.index.Indexed;
import org.springframework.data.mongodb.core.mapping.Document;

import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;

@Data
@NoArgsConstructor
@AllArgsConstructor
@Builder
@Document(collection = "leads")
@CompoundIndex(name = "lead_identifiers_index", def = "{'email': 1, 'phoneNumber': 1, 'aadharNumber': 1}")
@CompoundIndex(name = "lead_dashboard_index", def = "{'pId': 1, 'sourceId': 1, 'createdAt': -1}")
public class Lead {
    
    @Id
    private String id;
    
    @Indexed(unique = true)
    private String leadId;
    
    private String name;
    
    @Indexed
    private String phoneNumber;
    
    @Indexed
    private String email;
    
    @Indexed
    private String aadharNumber;
    
    private String sourceId;
    private String pId;
    
    @CreatedDate
    private LocalDateTime createdAt;
    
    // Merge metadata
    @Builder.Default
    private List<MergeRecord> mergedFrom = new ArrayList<>();
    
    @Builder.Default
    private List<String> sourcesSeen = new ArrayList<>();
    
    @Builder.Default
    private List<String> productsSeen = new ArrayList<>();
    
    // AI scoring stub fields
    private Integer leadScore;
    private String scoreReason;
    
    @LastModifiedDate
    private LocalDateTime updatedAt;
    
    @Data
    @NoArgsConstructor
    @AllArgsConstructor
    @Builder
    public static class MergeRecord {
        private LocalDateTime timestamp;
        private String sourceId;
        private String pId;
        private Object data; // Can store raw row data
    }
}
